### PR TITLE
Add opening times to GP display

### PR DESCRIPTION
**Pull request by neilbmclaughlin**
Still much refactoring to do but created pull request for visibility.

### CODE REVIEW

**Comment by neilbmclaughlin** [comment]
@st3v3nhunt could you take a look at this as I think it is close to being good enough to merge.

**Comment by st3v3nhunt** [comment]
Test notification @st3v3nhunt

**Comment by neilbmclaughlin** [comment]
Test notification @st3v3nhunt

File: app/middleware/gp.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+function getDetails(req, res, next) {
+  http.get(req.urlForGp, (response) => {
+    let syndicationXml = '';
+    response.on('data', (chunk) => {
+      syndicationXml += chunk;
+    });
+
+    response.on('end', () => {
+      if (response.statusCode === 200) {
+        /* Disabled since assigning to res is recommended best practice my Express */
```
Comment (st3v3nhunt) [inline app/middleware/gp.js]: Ultra picky: `my` should be `by`?
And can the wording be changed to `...assigning to the response is...`?

File: app/middleware/gp.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      }
+    });
+  }).on('error', (e) => {
+    console.log('Got an error: ', e);
+    next(e);
+  });
+}
+
+function getOpeningTimes(req, res, next) {
+  http.get(req.gpDetails.overviewLink, (response) => {
```
Comment (st3v3nhunt) [inline app/middleware/gp.js:37]: What happens if the link or `gpDetails` is null/empty?

File: app/middleware/gp.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    let syndicationXml = '';
+    response.on('data', (chunk) => {
+      syndicationXml += chunk;
+    });
+
+    response.on('end', () => {
+      if (response.statusCode === 200) {
+        /* Disabled since assigning to res is recommended best practice my Express */
```
Comment (st3v3nhunt) [inline app/middleware/gp.js]: Same as comment for line 16

File: app/middleware/gp.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        /* eslint-disable no-param-reassign */
+        req.gpDetails.openingTimes = {
+          reception: gpOpeningTimesParser('reception', syndicationXml),
+          surgery: gpOpeningTimesParser('surgery', syndicationXml),
+        };
+        // console.log(util.inspect(req.gpDetails, false, null));
```
Comment (st3v3nhunt) [inline app/middleware/gp.js]: Can this be removed if it is commented out?

File: app/utilities/gpDetailsParser.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+const parseGpDetailsFromSyndicationXml = (xml) => {
+  let gpDetails;
+  const options = {
+    tagNameProcessors: [stripPrefix],
+    ignoreAttrs: true,
+    explicitArray: false,
+  };
+  const xmlParser = new xml2js.Parser(options);
+  xmlParser.parseString(xml, (err, result) => {
+    // console.log(util.inspect(result, false, null));
```
Comment (st3v3nhunt) [inline app/utilities/gpDetailsParser.js]: Delete or uncomment?

File: test/unit/gpDetailsParser.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      expect(gpDetails.name).to.equal('A Ditri');
+      expect(gpDetails.address).to.eql(
+        {
+          line1: 'Chesser Surgery',
+          line2: '121 Wrythe Lane',
+          line3: 'Carshalton',
+          line4: 'Surrey',
+          postcode: 'SM5 2RT',
+        });
+      expect(gpDetails.overviewLink).to.equal('http://v1.syndication.nhschoices.nhs.uk/organisations/gppractices/14500/overview.xml?apikey=secret');
```
Comment (st3v3nhunt) [inline test/unit/gpDetailsParser.js:20]: Is this making a request to the real service?

File: test/unit/middleware.js [hunk 7]
```diff
@@ -23,9 +28,105 @@ describe('Middleware', () => {
...
+        done();
+      });
+    });
+    it('should handle http error', (done) => {
+      nock('http://test')
+        .get('/test')
+        .replyWithError('Bad HTTP stuff happened');
+
+      middleware.getDetails(fakeRequest, {}, (err) => {
+        expect(err.message).to.equal('Bad HTTP stuff happened');
```
Comment (st3v3nhunt) [inline test/unit/middleware.js]: Can this be replaced with a variable to make it clear it should be the same value as returned in line 75?

File: test/unit/middleware.js [hunk 8]
```diff
@@ -23,9 +28,105 @@ describe('Middleware', () => {
...
+      gpDetails: { overviewLink: 'http://test/test' },
+    };
+    it('should return populated opening times for the practice', (done) => {
+      nock('http://test')
+        .get('/test')
+        .reply(200, getSampleResponse('gp_overview'));
+
+      middleware.getOpeningTimes(fakeRequest, {}, () => {
+        expect(fakeRequest.gpDetails.openingTimes.reception).to.be.an('object');
+        expect(fakeRequest.gpDetails.openingTimes.reception).to.have.keys(daysOfTheWeek);
```
Comment (st3v3nhunt) [inline test/unit/middleware.js]: Do both the checks for the object and keys need to made? Could just the keys be checked?

File: test/unit/middleware.js [hunk 9]
```diff
@@ -23,9 +28,105 @@ describe('Middleware', () => {
...
+        done();
+      });
+    });
+    it('should handle http error', (done) => {
+      nock('http://test')
+        .get('/test')
+        .replyWithError('Bad HTTP stuff happened');
+
+      middleware.getOpeningTimes(fakeRequest, {}, (err) => {
+        expect(err.message).to.equal('Bad HTTP stuff happened');
```
Comment (st3v3nhunt) [inline test/unit/middleware.js]: Same as previous comment - can this be replaced with a variable for the error message along?

File: app/middleware/gp.js (hunk 2 above)
Comment (neilbmclaughlin) [inline app/middleware/gp.js:37]: I think this should be addressed as a general review of parameter validation and error handling in a separate piece of work to be done next. Will create an issue for this.

File: test/unit/gpDetailsParser.js (hunk 6 above)
Comment (neilbmclaughlin) [inline test/unit/gpDetailsParser.js:20]: No - getSampleResponse reads a xml document captured from the syndication service and stored in the test/resources/syndication_responses directory.

File: test/unit/middleware.js (hunk 7 above)
Comment (neilbmclaughlin) [inline test/unit/middleware.js]: I accept some level of duplication in tests in return for clarity and risk reduction. In the end 'Bad HTTP stuff happened' and badHttpStuffHappened are both testing that tokens are the same so are actually equivalent. My issue comes if I mistype the variable name as badHttpStuff in the expectation and the error message is contains a bug which means it is undefined then the test will pass. This is mitigated by always having a failing test first.

File: test/unit/middleware.js (hunk 9 above)
Comment (neilbmclaughlin) [inline test/unit/middleware.js]: See comments above.

File: app/middleware/gp.js (hunk 2 above)
Comment (st3v3nhunt) [inline app/middleware/gp.js:37]: Cool. Issue #72

File: test/unit/gpDetailsParser.js (hunk 6 above)
Comment (st3v3nhunt) [inline test/unit/gpDetailsParser.js:20]: Ah yeah, my mistake.

File: test/unit/middleware.js (hunk 7 above)
Comment (st3v3nhunt) [inline test/unit/middleware.js]: Mmh, not sure I fully understand or perhaps I do but I don't see it the same way. My take on it is...I don't really care what the error message is and don't want a typo to cause a test to fail. I want the test to fail if the error handling logic is incorrect/changes. Having the error message in a variable still means the test will be failing until such a time as the logic has been implemented. At least that is my understanding. Perhaps a voice/video conversation is required?

**Comment by neilbmclaughlin** [comment]
@st3v3nhunt I've committed the changes in response to the comments. Happy to talk about the string vs variable issue either before or after the merge.

**Pull request merged**